File: haskell/glue/src/Glue/Lib/Builtin/At.hs
module Glue.Lib.Builtin.At (at) where

import Data.List ((!?))
import Data.Map qualified as M
import Data.Maybe (fromMaybe)
import Data.Text qualified as T
import Glue.Eval (Eval, throwError)
import Glue.Eval.Exception (wrongArgumentType)
import Glue.IR (IR (..), Value (..))

-- | At function - univeral accessor
at :: IR Eval
at = NativeFunc \target ->
    pure $ NativeFunc \accessor ->
        case (target, accessor) of
            (String text, Integer index) ->
                pure
                    if index >= 0 && index < T.length text
                        then String . T.singleton $ T.index text index
                        else Void
            (List list, Integer index) ->
                pure . fromMaybe Void $ list !? index
            (Object object, String field) ->
                pure . fromMaybe Void $ M.lookup field object
            (Object object, Symbol field) ->
                pure . fromMaybe Void $ M.lookup field object
            (Object object, DottedSymbol field) -> nestedLookup field object
            (NativeValue (Value _ getters), String field) ->
                fromMaybe (pure Void) $ M.lookup field getters
            (NativeValue (Value _ getters), Symbol field) ->
                fromMaybe (pure Void) $ M.lookup field getters
            (NativeValue (Value _ getters), DottedSymbol fields) -> nestedNativeLookup fields getters
            _ -> throwError $ wrongArgumentType ["target", "accessor"]
  where
    nestedLookup [] _ = pure Void
    nestedLookup [key] obj =
        pure . fromMaybe Void $ M.lookup key obj
    nestedLookup (key : rest) obj =
        case M.lookup key obj of
            Just (Object subObj) -> nestedLookup rest subObj
            _ -> pure Void

    nestedNativeLookup [] _ = pure Void
    nestedNativeLookup [key] getters =
        fromMaybe (pure Void) $ M.lookup key getters
    nestedNativeLookup (key : rest) getters =
        case M.lookup key getters of
            Just evalAction -> do
                resolvedIR <- evalAction
                case resolvedIR of
                    Object subObj -> nestedLookup rest subObj
                    NativeValue (Value _ subGetters) -> nestedNativeLookup rest subGetters
                    _ -> pure Void
            _ -> pure Void

File: dart/glue/lib/src/lib/builtin/at.dart
import 'package:glue/error.dart';
import 'package:glue/eval.dart';
import 'package:glue/src/ir.dart';

final Ir at = IrNativeFunc((Ir target) {
  return Eval.pure(
    IrNativeFunc((Ir accessor) {
      switch (target) {
        case IrString(:final value) when accessor is IrInteger:
          final index = accessor.value;
          if (index >= 0 && index < value.length) {
            return Eval.pure(IrString(value[index]));
          }
          return Eval.pure(IrVoid());

        case IrList(:final elements) when accessor is IrInteger:
          final index = accessor.value;
          if (index >= 0 && index < elements.length) {
            return Eval.pure(elements[index]);
          }
          return Eval.pure(IrVoid());

        case IrObject(:final properties) when accessor is IrString:
          return Eval.pure(properties[accessor.value] ?? IrVoid());

        case IrObject(:final properties) when accessor is IrSymbol:
          return Eval.pure(properties[accessor.value] ?? IrVoid());

        case IrObject(:final properties) when accessor is IrDottedSymbol:
          return _nestedLookup(accessor.parts, properties.unlock);

        case IrNativeValue(:final value) when accessor is IrString:
          return value.getters[accessor.value] ?? Eval.pure(IrVoid());

        case IrNativeValue(:final value) when accessor is IrSymbol:
          return value.getters[accessor.value] ?? Eval.pure(IrVoid());

        case IrNativeValue(:final value) when accessor is IrDottedSymbol:
          return _nestedNativeLookup(accessor.parts, value.getters);

        default:
          return throwError(wrongArgumentType(['target', 'accessor']));
      }
    }),
  );
});

Eval<Ir> _nestedLookup(List<String> fields, Map<String, Ir> obj) =>
    switch (fields) {
      [] => Eval.pure(IrVoid()),
      [final key] => Eval.pure(obj[key] ?? IrVoid()),
      [final key, ...final rest] => switch (obj[key]) {
        IrObject(:final properties) => _nestedLookup(rest, properties.unlock),
        _ => Eval.pure(IrVoid()),
      },
    };

Eval<Ir> _nestedNativeLookup(
  List<String> fields,
  Map<String, Eval<Ir>> getters,
) => switch (fields) {
  [] => Eval.pure(IrVoid()),
  [final key] => getters[key] ?? Eval.pure(IrVoid()),
  [final key, ...final rest] => switch (getters[key]) {
    null => Eval.pure(IrVoid()),
    final evalAction => evalAction.bind(
      (resolvedIR) => switch (resolvedIR) {
        IrObject(:final properties) => _nestedLookup(rest, properties.unlock),
        IrNativeValue(:final value) => _nestedNativeLookup(rest, value.getters),
        _ => Eval.pure(IrVoid()),
      },
    ),
  },
};
